File: watch/YaverWatch/SessionClient.swift
// SessionClient.swift — standalone-mode transport that drives a LIVE coding
// session via POST /runner/session/turn.
//
// This is the endpoint a WATCH should call (docs/yaver-watch-surface.md §4.2).
// The older /watch/turn (AgentClient.swift) spawns a NEW task; this one drives
// the session the user already has running — "keep developing this" means the
// ubuntu session, not a fresh task.
//
// Maps the session endpoint's response (awaitingChoice + options[] + pane) into
// the SAME WatchReply the watch already renders:
//   awaitingChoice: true  → confirmNeeded (prompt = options list, token = sentinel)
//   awaitingChoice: false → summary (spoken = first clean line of pane)
//   error / unreachable    → error
//
// Choice loop: the client tracks `lastAwaitingChoice`. When the user speaks a
// number after a menu, sendText() recognizes it as a choice and sends {choice}
// instead of {text}. ConfirmView's confirm/cancel maps to choice "1"/"2" as a
// fallback (voice is the preferred path for picking a specific option).
//
// The four server-side guards (docs/yaver-watch-surface.md §4.2) are honored
// by the endpoint itself — this client just maps the response. Never type a
// prompt into a menu, never send a choice without a menu, never append Enter
// to a digit, always re-read after settling. The 409s this client may receive
// are the guards firing; they map to confirm-needed or error, not a retry.

import Foundation

/// Sentinel token for session-choice confirms. Distinguishes a confirm-needed
/// that came from the session endpoint (awaitingChoice) from one that came from
/// the phone's risk gate. When the watch sees this token, ConfirmView's
/// confirm/cancel maps to choice "1"/"2" instead of a transcript echo.
let sessionChoiceToken = "__yaver_session_choice__"

actor SessionClient {
    private let token: String
    private let box: BoxTarget
    private let session: URLSession

    /// Tracks whether the last response from the box had `awaitingChoice: true`.
    /// When the user speaks a number after a menu, sendText() uses this to send
    /// `{choice}` instead of `{text}` — the user doesn't have to know the
    /// protocol changed under them.
    private var lastAwaitingChoice = false

    init(token: String, box: BoxTarget) {
        self.token = token
        self.box = box
        let cfg = URLSessionConfiguration.default
        // The session endpoint waits up to `waitMs` for the runner to react,
        // then reads the pane. 30s headroom covers the 6s default wait + settle.
        cfg.timeoutIntervalForRequest = 30
        self.session = URLSession(configuration: cfg)
    }

    // MARK: - Public API (mirrors AgentClient's turn surface)

    /// Send a spoken transcript. If the box was awaiting a choice and the text
    /// looks like a number or number-word, send it as `{choice}` instead.
    func sendText(_ text: String) async throws -> WatchReply {
        if lastAwaitingChoice, let choice = SessionClient.parseChoice(text) {
            return try await turn(text: nil, choice: choice)
        }
        return try await turn(text: text, choice: nil)
    }

    /// Send a menu choice directly (e.g. from ConfirmView confirm → "1").
    func sendChoice(_ choice: String) async throws -> WatchReply {
        try await turn(text: nil, choice: choice)
    }

    /// Map a ConfirmReply (from ConfirmView) to a session choice.
    /// confirm → option "1" (first option), cancel → option "2" (second).
    /// This is a lossy fallback — the voice path (speak the number) is preferred
    /// because menus renumber and option 1 isn't always "yes".
    func sendConfirm(reply: ConfirmReply) async throws -> WatchReply {
        let choice = reply == .confirm ? "1" : "2"
        return try await turn(text: nil, choice: choice)
    }

    // MARK: - Core POST

    private struct SessionResponse: Decodable {
        let ok: Bool?
        let session: String?
        let runner: String?
        let sent: String?
        let awaitingChoice: Bool?
        let options: [String]?
        let pane: String?
        let error: String?
    }

    private func turn(text: String?, choice: String?) async throws -> WatchReply {
        guard let url = URL(string: "http://\(box.host):\(box.port)/runner/session/turn") else {
            throw AgentError(message: "bad box host")
        }
        var body: [String: Any] = [
            "waitMs": 6000, // short + snappy for a wrist; the agent default
        ]
        if let text { body["text"] = text }
        if let choice { body["choice"] = choice }

        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        req.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, resp) = try await session.data(for: req)
        guard let http = resp as? HTTPURLResponse else {
            throw AgentError(message: "no response")
        }

        // 409 is the guards firing — the pane is showing a menu (can't take a
        // prompt) or isn't showing one (can't take a choice). The response body
        // still carries awaitingChoice + options + pane, so decode and map it
        // the same way as a 200.
        guard (200..<300).contains(http.statusCode) || http.statusCode == 409 else {
            if let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
               let err = obj["error"] as? String {
                throw AgentError(message: err)
            }
            throw AgentError(message: "session turn failed (\(http.statusCode))")
        }

        let sr: SessionResponse
        do {
            sr = try JSONDecoder().decode(SessionResponse.self, from: data)
        } catch {
            throw AgentError(message: "bad response from box")
        }

        // Track whether the pane is on a menu so the next sendText can map a
        // spoken number to a choice automatically.
        lastAwaitingChoice = sr.awaitingChoice ?? false

        // Map the session response → WatchReply (same kinds the watch renders).
        if sr.awaitingChoice == true {
            // The pane is showing a menu. Show the options as the prompt; the
            // user speaks a number (or taps Confirm for option 1).
            let options = sr.options ?? []
            let prompt = options.isEmpty
                ? "Choose an option."
                : options.joined(separator: "\n")
            return WatchReply(
                kind: .confirmNeeded,
                spoken: options.isEmpty ? "Pick an option." : SessionClient.speakOptions(options),
                token: sessionChoiceToken,
                prompt: prompt
            )
        }

        // Not awaiting a choice → summarize the pane to one sentence.
        let pane = sr.pane ?? ""
        let spoken = SessionClient.summarize(pane)
        if sr.ok == false || (sr.error?.isEmpty == false && spoken.isEmpty) {
            return WatchReply(kind: .error, spoken: sr.error ?? "Something went wrong.")
        }
        return WatchReply(kind: .summary, spoken: spoken)
    }

    // MARK: - Pane summarization (mirrors watch_risk.go::watchFirstStatusClause)

    /// Pull the first short, status-shaped clause out of a pane tail. Refuses
    /// code/markup/path-dump lines (the watch must never speak code). Clamps to
    /// 120 chars — the same ceiling the agent's watch summariser uses.
    private static let codePattern = try! NSRegularExpression(
        pattern: #"[{}<>;=]|```|\b(function|const|class|def|import|return)\b|/\w+/"#
    )
    private static let sentencePattern = try! NSRegularExpression(
        pattern: #"^(.{1,120}?[.!?])(\s|$)"#
    )
    private static let markdownPattern = try! NSRegularExpression(
        pattern: "[#*`_~]"
    )

    static func summarize(_ pane: String) -> String {
        let lines = pane.split(separator: "\n", omittingEmptySubsequences: true)
        guard let first = lines.first else { return "Done." }
        let firstLine = String(first).trimmingCharacters(in: .whitespaces)
        if firstLine.isEmpty { return "Done." }

        // Refuse code/markup/path-dump shaped lines.
        let range = NSRange(firstLine.startIndex..., in: firstLine)
        if codePattern.firstMatch(in: firstLine, range: range) != nil {
            // Fall through to the next clean line if any, else default.
            for line in lines.dropFirst() {
                let l = String(line).trimmingCharacters(in: .whitespaces)
                if l.isEmpty { continue }
                let r = NSRange(l.startIndex..., in: l)
                if codePattern.firstMatch(in: l, range: r) == nil {
                    return clampSentence(stripMarkdown(l))
                }
            }
            return "Done."
        }

        return clampSentence(stripMarkdown(firstLine))
    }

    private static func clampSentence(_ s: String) -> String {
        // First sentence only.
        let r = NSRange(s.startIndex..., in: s)
        if let m = sentencePattern.firstMatch(in: s, range: r),
           let sentenceRange = Range(m.range(at: 1), in: s) {
            let clause = String(s[sentenceRange])
            return clause.count <= 120 ? clause : String(clause.prefix(119)) + "…"
        }
        // No sentence terminator — clamp the whole line.
        if s.count <= 120 { return s }
        return String(s.prefix(119)) + "…"
    }

    private static func stripMarkdown(_ s: String) -> String {
        let r = NSRange(s.startIndex..., in: s)
        return markdownPattern.stringByReplacingMatches(
            in: s, range: r, withTemplate: ""
        ).trimmingCharacters(in: .whitespaces)
    }

    /// Speak the options as a short voice prompt: "Choose: 1. Yes. 2. No."
    /// Clamped so a long menu doesn't drone on — the user can glance at the
    /// screen for the full list.
    private static func speakOptions(_ options: [String]) -> String {
        let short = options.prefix(4).map { opt in
            // "1. Yes, I trust this folder" → "1. Yes I trust this folder"
            // (strip the comma-clause after 40 chars for speech brevity)
            let trimmed = opt.trimmingCharacters(in: .whitespaces)
            if trimmed.count <= 60 { return trimmed }
            return String(trimmed.prefix(59)) + "…"
        }.joined(separator: ". ")
        return "Choose: \(short)."
    }

    /// Map a spoken word/number to a bare digit string the session endpoint
    /// accepts (`isTmuxChoiceAnswer` = `^\s*\d{1,2}\s*$`).
    private static func parseChoice(_ text: String) -> String? {
        let t = text.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        if t.isEmpty { return nil }
        // Bare digit ("1", "2", "12").
        if t.allSatisfy(\.isNumber) { return t }
        // Common number words.
        let wordMap: [String: String] = [
            "one": "1", "two": "2", "three": "3", "four": "4", "five": "5",
            "six": "6", "seven": "7", "eight": "8", "nine": "9",
            "first": "1", "second": "2", "third": "3",
            "yes": "1", "no": "2", // common menu shapes
        ]
        return wordMap[t]
    }
}

File: wear/app/src/main/kotlin/io/yaver/wear/SessionClient.kt
package io.yaver.wear

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject
import java.util.concurrent.TimeUnit

/**
 * Standalone transport that drives a LIVE coding session via
 * POST /runner/session/turn.
 *
 * This is the endpoint a WATCH should call (docs/yaver-watch-surface.md §4.2).
 * The older /watch/turn (AgentClient.kt) spawns a NEW task; this one drives the
 * session the user already has running — "keep developing this" means the ubuntu
 * session, not a fresh task.
 *
 * Maps the session endpoint's response (awaitingChoice + options[] + pane) into
 * the SAME [WatchProtocol.Reply] the watch already renders:
 *   awaitingChoice: true  → [WatchProtocol.Reply.ConfirmNeeded]
 *   awaitingChoice: false → [WatchProtocol.Reply.Summary] (spoken = first clean line of pane)
 *   error / unreachable    → [WatchProtocol.Reply.Error]
 *
 * Choice loop: the client tracks `lastAwaitingChoice`. When the user speaks a
 * number after a menu, [sendText] recognizes it as a choice and sends {choice}
 * instead of {text}. ConfirmScreen's confirm/cancel maps to choice "1"/"2" as a
 * fallback (voice is the preferred path for picking a specific option).
 *
 * The four server-side guards (docs/yaver-watch-surface.md §4.2) are honored
 * by the endpoint itself — this client just maps the response. 409 status codes
 * (the guards firing) are decoded the same as 200: the body still carries
 * awaitingChoice + options + pane.
 *
 * Never throws — network/parse failures become [WatchProtocol.Reply.Error] so
 * the wrist always shows a line (same pattern as AgentClient).
 */
class SessionClient(
    /** e.g. "http://192.168.1.50:18080" — host:port of the box on the LAN. */
    private val boxBaseUrl: String,
    private val bearerToken: String,
) {

    private val http: OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(5, TimeUnit.SECONDS)
        // The session endpoint waits up to `waitMs` for the runner to react,
        // then reads the pane. 30s covers the 6s default wait + settle.
        .readTimeout(30, TimeUnit.SECONDS)
        .build()

    /** Tracks whether the last response had awaitingChoice: true. When the user
     *  speaks a number after a menu, sendText() uses this to send {choice}. */
    @Volatile
    private var lastAwaitingChoice: Boolean = false

    // ---- Public API ------------------------------------------------------

    /**
     * Send a spoken transcript. If the box was awaiting a choice and the text
     * looks like a number or number-word, send it as {choice} instead.
     */
    suspend fun sendText(text: String): WatchProtocol.Reply = withContext(Dispatchers.IO) {
        if (lastAwaitingChoice) {
            parseChoice(text)?.let { choice -> return@withContext turn(text = null, choice = choice) }
        }
        turn(text = text, choice = null)
    }

    /** Send a menu choice directly (e.g. from ConfirmScreen confirm → "1"). */
    suspend fun sendChoice(choice: String): WatchProtocol.Reply =
        withContext(Dispatchers.IO) { turn(text = null, choice = choice) }

    /**
     * Map a [WatchProtocol.ConfirmReply] (from ConfirmScreen) to a session choice.
     * confirm → option "1" (first option), cancel → option "2" (second).
     * This is a lossy fallback — the voice path (speak the number) is preferred
     * because menus renumber and option 1 isn't always "yes".
     */
    suspend fun sendConfirm(reply: WatchProtocol.ConfirmReply): WatchProtocol.Reply =
        withContext(Dispatchers.IO) {
            val choice = if (reply == WatchProtocol.ConfirmReply.CONFIRM) "1" else "2"
            turn(text = null, choice = choice)
        }

    // ---- Core POST -------------------------------------------------------

    private suspend fun turn(text: String?, choice: String?): WatchProtocol.Reply {
        return try {
            val body = JSONObject().apply {
                put("waitMs", 6000) // short + snappy for a wrist
                if (text != null) put("text", text)
                if (choice != null) put("choice", choice)
            }.toString().toRequestBody(JSON)

            val request = Request.Builder()
                .url(boxBaseUrl.trimEnd('/') + "/runner/session/turn")
                .header("Authorization", "Bearer $bearerToken")
                .post(body)
                .build()

            http.newCall(request).execute().use { resp ->
                val text = resp.body?.string().orEmpty()
                // 409 is the guards firing — the body still carries
                // awaitingChoice + options + pane, so decode it the same as 200.
                if (!resp.isSuccessful && resp.code != 409) {
                    val err = try {
                        JSONObject(text).optString("error", "")
                    } catch (_: Throwable) { "" }
                    return@use WatchProtocol.Reply.Error(
                        if (err.isNotEmpty()) err
                        else when (resp.code) {
                            401 -> "Sign in again."
                            in 500..599 -> "Your box hit an error."
                            else -> "I couldn't reach your box."
                        }
                    )
                }
                mapSessionResponse(text)
            }
        } catch (_: Throwable) {
            // Connection refused / timeout — the box is unreachable (likely a
            // self-parked managed box). Flag it so the wrist offers "Wake".
            WatchProtocol.Reply.Error("I couldn't reach your box.", boxUnreachable = true)
        }
    }

    /** Map the session endpoint's JSON response → WatchProtocol.Reply. */
    private fun mapSessionResponse(json: String): WatchProtocol.Reply {
        val obj = try {
            JSONObject(json)
        } catch (_: Throwable) {
            return WatchProtocol.Reply.Error("Something went wrong.")
        }

        val awaiting = obj.optBoolean("awaitingChoice", false)
        lastAwaitingChoice = awaiting

        if (awaiting) {
            // The pane is showing a menu. Show the options as the prompt; the
            // user speaks a number (or taps Confirm for option 1).
            // The prompt IS what gets spoken (Speech.forReply speaks it for
            // ConfirmNeeded) and shown on ConfirmScreen — both paths read the
            // same field, so the options list serves double duty.
            val options = obj.optJSONArray("options")
            val optionList = options?.toStringList() ?: emptyList()
            val prompt = if (optionList.isEmpty()) "Choose an option."
                         else optionList.joinToString("\n")
            return WatchProtocol.Reply.ConfirmNeeded(
                token = SESSION_CHOICE_TOKEN,
                prompt = prompt,
            )
        }

        // Not awaiting a choice → summarize the pane to one sentence.
        val pane = obj.optString("pane", "")
        val ok = obj.optBoolean("ok", true)
        val error = obj.optString("error", "")
        val spoken = summarizePane(pane)

        if (!ok || (error.isNotEmpty() && spoken.isEmpty())) {
            return WatchProtocol.Reply.Error(error.ifEmpty { "Something went wrong." })
        }
        val session = obj.optString("session", "")
        return WatchProtocol.Reply.Summary(
            taskId = session,
            status = "completed",
            spoken = spoken,
        )
    }

    // ---- Pane summarization (mirrors watch_risk.go::watchFirstStatusClause) ----

    /**
     * Pull the first short, status-shaped clause out of a pane tail. Refuses
     * code/markup/path-dump lines (the watch must never speak code). Clamps to
     * 120 chars — the same ceiling the agent's watch summariser uses.
     */
    private fun summarizePane(pane: String): String {
        val lines = pane.split("\n").map { it.trim() }.filter { it.isNotEmpty() }
        if (lines.isEmpty()) return "Done."

        for (line in lines) {
            if (!looksLikeCode(line)) {
                return clampSentence(stripMarkdown(line))
            }
        }
        // All lines looked like code.
        return "Done."
    }

    /** True when a line smells like code/markup/path-dump (mirrors the Go regex). */
    private fun looksLikeCode(line: String): Boolean {
        if (CODE_PATTERN.containsMatchIn(line)) return true
        if (line.contains("```")) return true
        return false
    }

    /** First sentence only, clamped to 120 chars. */
    private fun clampSentence(s: String): String {
        val m = SENTENCE_PATTERN.find(s)
        if (m != null) {
            val clause = m.groupValues[1]
            return if (clause.length <= 120) clause else clause.take(119) + "…"
        }
        return if (s.length <= 120) s else s.take(119) + "…"
    }

    private fun stripMarkdown(s: String): String =
        MARKDOWN_PATTERN.replace(s, "").trim()

    /** Map a spoken word/number to a bare digit the session endpoint accepts. */
    private fun parseChoice(text: String): String? {
        val t = text.trim().lowercase()
        if (t.isEmpty()) return null
        // Bare digit ("1", "2", "12").
        if (t.all { it.isDigit() }) return t
        // Common number words.
        return WORD_TO_NUMBER[t]
    }

    /** Helper: JSONArray → List<String> (null-safe). */
    private fun JSONArray.toStringList(): List<String> =
        (0 until length()).mapNotNull { i -> optString(i, "").takeIf { it.isNotEmpty() } }

    companion object {
        private val JSON = "application/json; charset=utf-8".toMediaType()

        /** Sentinel token for session-choice confirms (mirrors Swift). */
        const val SESSION_CHOICE_TOKEN = "__yaver_session_choice__"

        // Mirrors watch_risk.go::watchFirstStatusClause code-detection regex.
        private val CODE_PATTERN = Regex(
            """[{}<>;=]|\b(function|const|class|def|import|return)\b|/\w+/"""
        )
        private val SENTENCE_PATTERN = Regex("""^(.{1,120}?[.!?])(\s|$)""")
        private val MARKDOWN_PATTERN = Regex("[#*`_~]")

        private val WORD_TO_NUMBER = mapOf(
            "one" to "1", "two" to "2", "three" to "3", "four" to "4",
            "five" to "5", "six" to "6", "seven" to "7", "eight" to "8",
            "nine" to "9", "first" to "1", "second" to "2", "third" to "3",
            "yes" to "1", "no" to "2",
        )
    }
}
